### PR TITLE
rename PV during restore when cloning a namespace if PV already exists in-cluster

### LINKED ISSUE #192: Support restoring a PV into the same cluster
Forked from #189 

Scenario:
- PVC in namespace _N1_, bound to PV _P1_
- Create backup (snapshotting the PV)
- Restore backup, remapping namespace _N1_ to _N2_
- Have the PVC in _N2_ reference a **new** PV _P2_, created from _P1_'s snapshot

Currently, this is not possible, because Ark does not change the name of the PV when restoring it. The result is 2 PVCs both trying to claim the same (original) PV _P1_.

Questions:
- Should this be done automatically?
- Should this be optional?
- What sort of controls/flags should we expose to the user?
- How do we cover the 80-90% use case?

cc @evaldasou

### PR DESCRIPTION
**Pull request by skriss**
Signed-off-by: Steve Kriss <krisss@vmware.com>

closes #192

TODO:
- [ ] add unit test cases (positive cases & ensure adequate negative cases)
- [ ] think about possible ways to simplify this code

### CODE REVIEW

File: pkg/restore/restore.go [hunk 1]
```diff
@@ -876,15 +879,32 @@ func (ctx *context) restoreItem(obj *unstructured.Unstructured, groupResource sc
...
-			shouldRestoreSnapshot, err := ctx.shouldRestore(name, resourceClient)
+			renamePV, err := renamePV(ctx, obj, resourceClient)
 			if err != nil {
-				addToResult(&errs, namespace, errors.Wrapf(err, "error waiting on in-cluster persistentvolume %s", name))
+				addToResult(&errs, namespace, err)
 				return warnings, errs
 			}
 
+			var shouldRestoreSnapshot bool
+			switch renamePV {
```
Comment (prydonius) [inline pkg/restore/restore.go]: any reason to use switch over if/else here?

File: pkg/restore/restore.go [hunk 2]
```diff
@@ -1103,6 +1148,45 @@ func (ctx *context) restoreItem(obj *unstructured.Unstructured, groupResource sc
 	return warnings, errs
 }
 
+// renamePV returns a boolean indicating whether a persistent volume should be given a new name
+// before being restored, or an error if this cannot be determined. A persistent volume will be
+// given a new name if and only if (a) a PV with the original name already exists in-cluster, and
+// (b) in the backup, the PV is claimed by a PVC in a namespace that's being remapped during the
+// restore.
+func renamePV(ctx *context, obj *unstructured.Unstructured, client client.Dynamic) (bool, error) {
```
Comment (prydonius) [inline pkg/restore/restore.go]: small nitpick: we might want to call this `shouldRenamePV` to make it clearer that this call doesn't do the renaming

File: pkg/restore/restore.go (hunk 1 above)
Comment (skriss) [reply pkg/restore/restore.go]: not really, i think this started out being more complicated and then got simplified. switching.

**Comment by skriss** [comment]
Addressed comments so far + added test cases

File: pkg/restore/restore.go [hunk 3]
```diff
@@ -876,15 +883,31 @@ func (ctx *context) restoreItem(obj *unstructured.Unstructured, groupResource sc
...
 				return warnings, errs
 			}
 
+			var shouldRestoreSnapshot bool
+			if shouldRenamePV {
+				// if we're renaming the PV, we're going to give it a new random name,
+				// so we can assume it doesn't already exist in the cluster and therefore
+				// we should proceed with restoring from snapshot.
+				shouldRestoreSnapshot = true
+			} else {
```
Comment (carlisia) [inline pkg/restore/restore.go]: You could delete the `if` statement up there and instead of the `else` here do: `if ! shouldRenamePV`

File: pkg/restore/restore.go [hunk 4]
```diff
@@ -876,15 +883,31 @@ func (ctx *context) restoreItem(obj *unstructured.Unstructured, groupResource sc
...
+				// Check if the PV exists in the cluster before attempting to create
+				// a volume from the snapshot, in order to avoid orphaned volumes (GH #609)
+				shouldRestoreSnapshot, err = ctx.shouldRestore(name, resourceClient)
+				if err != nil {
+					addToResult(&errs, namespace, errors.Wrapf(err, "error waiting on in-cluster persistentvolume %s", name))
+					return warnings, errs
+				}
+			}
+
 			if shouldRestoreSnapshot {
```
Comment (carlisia) [inline pkg/restore/restore.go]: If you make the suggested change above, you could move the code from within this `if` to inside the `if ! shouldRenamePV`, right after the return on error.

File: pkg/restore/restore.go [hunk 5]
```diff
@@ -893,6 +916,23 @@ func (ctx *context) restoreItem(obj *unstructured.Unstructured, groupResource sc
 				}
 				obj = updatedObj
 			}
+
+			if shouldRenamePV {
```
Comment (carlisia) [inline pkg/restore/restore.go]: If you make both suggested changes above, then this could be an `else` to the `if !shouldRenamePV`. I think this would work.

File: pkg/restore/restore.go [hunk 6]
```diff
@@ -893,6 +916,23 @@ func (ctx *context) restoreItem(obj *unstructured.Unstructured, groupResource sc
...
+				obj.SetName(newName)
+
+				// add the original PV name as an annotation
+				annotations := obj.GetAnnotations()
+				if annotations == nil {
+					annotations = map[string]string{}
+				}
+				annotations["velero.io/original-pv-name"] = oldName
+				obj.SetAnnotations(annotations)
+			}
```
Comment (carlisia) [inline pkg/restore/restore.go]: Add a line after this bracket.

File: pkg/restore/restore.go (hunk 4 above)
Comment (skriss) [reply pkg/restore/restore.go]: I might be missing something, but I don't think this can move into the `if !shouldRenamePV { ... }` because it needs to be executed whether `shouldRenamePV` is true or not - `shouldRestoreSnapshot` can potentially be `true` in both cases.
